### ISSUE TITLE
Distinct on separate line causes internal error

### ISSUE THREAD

**Issue by holly-evans** [sqlfluff/sqlfluff#5553]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

I ran sqlfluff fix on this query and it failed with an internal error.
```sql
select
distinct
lead_id
from {{ ref('lead') }}
```

I ran sqlfluff lint on the same query and this was the output.
```
=== [dbt templater] Sorting Nodes...
=== [dbt templater] Compiling dbt project...
=== [dbt templater] Project Compiled.
== [projects/ncsa_dbt/models/marts/lint_me.sql] FAIL                                                                                                                                                         
L:   1 | P:   1 | LT09 | Select targets should be on a new line unless there is
                       | only one select target.
                       | [layout.select_targets]
L:   1 | P:   1 | LT10 | 'SELECT' modifiers (e.g. 'DISTINCT') must be on the same
                       | line as 'SELECT'. [layout.select_modifiers]
L:   3 | P:   1 | LT02 | Expected indent of 1 tabs. [layout.indent]
All Finished 📜 🎉!
```

I believe the error arises only when both LT09 and LT10 are violated. These similar queries were successfully linted.
```sql
select
distinct lead_id
from {{ ref('lead') }}

select distinct
lead_id
from {{ ref('lead') }}

select
distinct
lead_id,
created_at
from {{ ref('lead') }}
```

### Expected Behaviour

I expected SQLFluff to successfully fix the query to this.
```sql
select distinct lead_id
from {{ ref('lead') }}
```

### Observed Behaviour

SQLFluff could not complete the lint and experienced an internal error.

```
==== finding fixable violations ====
FORCE MODE: Attempting fixes...
=== [dbt templater] Sorting Nodes...
=== [dbt templater] Compiling dbt project...
=== [dbt templater] Project Compiled.
WARNING    Unable to lint projects/ncsa_dbt/models/marts/lint_me.sql due to an internal error. Please report this as an issue with your query's contents and stacktrace below!                               
To hide this warning, add the failing file to .sqlfluffignore
Traceback (most recent call last):
  File "/Users/hevans/.local/share/virtualenvs/data_dbt-AeL2xOvX/lib/python3.10/site-packages/sqlfluff/core/linter/runner.py", line 109, in run
    yield partial()
  File "/Users/hevans/.local/share/virtualenvs/data_dbt-AeL2xOvX/lib/python3.10/site-packages/sqlfluff/core/linter/linter.py", line 672, in lint_rendered
    return cls.lint_parsed(
  File "/Users/hevans/.local/share/virtualenvs/data_dbt-AeL2xOvX/lib/python3.10/site-packages/sqlfluff/core/linter/linter.py", line 592, in lint_parsed
    ) = cls.lint_fix_parsed(
  File "/Users/hevans/.local/share/virtualenvs/data_dbt-AeL2xOvX/lib/python3.10/site-packages/sqlfluff/core/linter/linter.py", line 498, in lint_fix_parsed
    new_tree, _, _ = tree.apply_fixes(
  File "/Users/hevans/.local/share/virtualenvs/data_dbt-AeL2xOvX/lib/python3.10/site-packages/sqlfluff/core/parser/segments/base.py", line 1494, in apply_fixes
    s, before, after = seg.apply_fixes(dialect, rule_code, fixes)
  File "/Users/hevans/.local/share/virtualenvs/data_dbt-AeL2xOvX/lib/python3.10/site-packages/sqlfluff/core/parser/segments/base.py", line 1494, in apply_fixes
    s, before, after = seg.apply_fixes(dialect, rule_code, fixes)
  File "/Users/hevans/.local/share/virtualenvs/data_dbt-AeL2xOvX/lib/python3.10/site-packages/sqlfluff/core/parser/segments/base.py", line 1494, in apply_fixes
    s, before, after = seg.apply_fixes(dialect, rule_code, fixes)
  File "/Users/hevans/.local/share/virtualenvs/data_dbt-AeL2xOvX/lib/python3.10/site-packages/sqlfluff/core/parser/segments/base.py", line 1511, in apply_fixes
    assert not self._find_start_or_end_non_code(seg_buffer), (
AssertionError: Found inappropriate fix application: inappropriate whitespace positioning. Post `_choose_anchor_segment`. Please report this issue on GitHub with your SQL query. 
```

### How to reproduce

Create `lint_me.sql` containing

```sql
select
distinct
lead_id
from {{ ref('lead') }}
```

Run `sqlfluff fix lint_me.sql`.

### Dialect

dialect = redshift
templater = dbt

### Version

sqlfluff==2.3.0
sqlfluff-templater-dbt==2.3.0
dbt-redshift==1.6.0

### Configuration

.sqlfluff
---------------
```toml
[sqlfluff]
dialect = redshift
templater = dbt
rules = core, LT09, RF06, CV02, LT11
exclude_rules = LT05, L016
large_file_skip_byte_limit = 30000

[sqlfluff:templater:dbt]
project_dir = projects/ncsa_dbt
profiles_dir = projects/ncsa_dbt
profile = redshift
target = docs

[sqlfluff:indentation]
indent_unit = tab
indented_ctes = True

[sqlfluff:layout:type:comma]
spacing_before = touch
line_position = trailing

[sqlfluff:rules:capitalisation.keywords]
capitalisation_policy = lower

[sqlfluff:rules:capitalisation.identifiers]
capitalisation_policy = lower

[sqlfluff:rules:capitalisation.functions]
extended_capitalisation_policy = lower

[sqlfluff:rules:capitalisation.literals]
capitalisation_policy = lower

[sqlfluff:rules:capitalisation.types]
extended_capitalisation_policy = lower
```

.sqlfluff
---------------
```toml
**/target
**/dbt_packages
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by peterbud** [comment]
With the latest `sqlfluff` the linters still sees both LT09 and LT10 rule violation, but properly fixes the sql.

```bash
sqlfluff lint test.sql -d redshift
== [test.sql] FAIL                                                                                                             
L:   1 | P:   1 | LT09 | Select targets should be on a new line unless there is
                       | only one select target. [layout.select_targets]
L:   1 | P:   1 | LT10 | 'SELECT' modifiers (e.g. 'DISTINCT') must be on the same
                       | line as 'SELECT'. [layout.select_modifiers]
L:   3 | P:   1 | LT02 | Expected indent of 4 spaces. [layout.indent]
L:   4 | P:  23 | LT12 | Files must end with a single trailing newline.
                       | [layout.end_of_file]
All Finished 📜 🎉!
```

Fixing:
```bash
sqlfluff fix test.sql -d redshift 
==== finding fixable violations ====
== [test.sql] FAIL                                                                                                             
L:   1 | P:   1 | LT09 | Select targets should be on a new line unless there is
                       | only one select target. [layout.select_targets]
L:   3 | P:   1 | LT02 | Expected indent of 4 spaces. [layout.indent]
L:   4 | P:  23 | LT12 | Files must end with a single trailing newline.
                       | [layout.end_of_file]
== [test.sql] FIXED
3 fixable linting violations found
```

Result
```sql
select distinct lead_id
from {{ ref('lead') }}
```
